ops/__init__.py

```diff
@@ -39,14 +39,129 @@
 Full developer documentation is available at https://juju.is/docs/sdk.
 """
 
+# The "from .X import Y" imports below don't explicitly tell Pyright (or MyPy)
+# that those symbols are part of the public API, so we have to add __all__.
+__all__ = [
```

Comment:
In discussing this with Ben, he pointed me towards:
https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md#library-interface

Which says that the other way of getting PyRight to consider these as Public interfaces is to do:
```
from .charm import (
  ActionEvent as ActionEvent
  ...
  )
```
We discussed whether this action of creating a local type alias is "more obvious" why you are doing it. And arguably it isn't a lot more obvious (but we have a caveat comment here to explain why we are duplicating `__all__`, so it isn't any *more* obvious, either).
However, the really big win is that ActionEvent as ActionEvent is 'local' and so when you want to add a new type that we have added to ops, you add 1 import line with both pieces and then it just works, rather than knowing that you need to import it in one place, and then reference that in all in another place. So it feels much easier to maintain.

Reply:
I actually agree with John's rationale here, and had a [patch](https://gist.github.com/benhoyt/9de89c7b67be250444d1866b43231df8) all ready to go -- it solves the issue with the type checking. However, Sphinx (our API docs generator), specifically its autodoc extension, doesn't pick up any names from `ops` as it still doesn't see those names as public.

You can have yet another explicit list of names in `docs/index.rst`, but that is error-prone and could easily get out of sync. There's probably a way to hack the Sphinx extension to be cleverer about this, but I've played around with it a bunch, and it doesn't seem to be easily configurable. So I don't think it's worth it.

Reverting back to using `__all__` explicitly, which Sphinx understands fine.

Reply:
A couple of other points regarding maintainability (in favour of `__all__`):

* With the `import X as X` approach, it would be easy to misspell the alias without noticing, like `from .model import Application as Appliction`.
* If you accidentally forget to add a name from the import list to the `__all__` list, Pyright complains (that "X is imported but not used"). If you accidentally have an additional name in the `__all__` list, again Pyright complains ("X is not defined"). This means CI will catch it if you add it in one place but not the other.